### ISSUE TITLE
Typed negative (non-existing) edge sampling for gnn methods

### ISSUE THREAD

**Issue by hosseinfani** [fani-lab/OpeNTF#348]
### For test, 

We predict the edge by dot product of member and team nodes' embeddings for pairs of member-team nodes. We build test set of edges from ALL member nodes to test team nodes. True positive (existing) edges are those members in the team and the rest experts are negative edges.  

> So, being member-to-team, or team-rev_to-member don't matter. A single dot product.

> But both types of edges are removed from the graph when passes to cross-fold loop of training-valid

> The actual test step is via opentf pipeline, i.e., loading the final model and asking it to predict the pos/neg edges using the nodes' embeddings of ALL members and the test team nodes

### For valid, 

Similar to test, we predict the edge by dot product of member and team nodes' embeddings for pairs of member-team nodes in validation set of teams. We build valid set of edges from true positive (existing) edges for those members in the valid teams and the **a selection of** rest of experts are negative edges. If we use ALL scenario as in test, the training loop over batches and epochs become very slow.   

> But both types of edges, i.e., member-to-team and team-rev_to-member, for positive edges are removed from the graph when passes to cross-fold loop of training-valid

> We ignore the supervision_edge_types config since it should not impact how validation set of team works

> For selection of negative edges, we use the ns config, i.e., ns times of positive edge we select from ALL members but not in valid teams.

> To calcuate valid loss both types of member-to-team, or team-rev_to-member are involved but but no issue => twice dot product but then averaged

### During training, 

to generate loss and update the node embeddings, i.e., supervision signal:

> If supervision_edge_types is None:

>>  Positive Edges: both types of edges are considered for 'to' and 'rev_to' for positive edges by the LinkNeighborLoader
>>> among all "existing" edge types even like skill-to-team, team-rev_to-skill
>>> the loss is still based on dot product of node embeddings, so the only effect is repeated loss between same node pairs in a batch, but at the end, the loss is averaged. no issue

>>  Negative Edges: Both types of edges are considered for 'to' and 'rev_to' for negative edges by our implemented function _sample_negatives_from_contexts (tnx to claude)
 >>> among all "non-existing" edge types even like skill-to-skill, team-to-team
 >>>  also we removed valid/test edges from the list as they should stay positive edges. Edges in test and valid sets, in "both" directions, should be removed from negative edges: both member to team and team rev_to member

> If supervision_edge_types is not None:
>> only among the supervision_edge_types, for positive edges by the LinkNeighborLoader
>> only among the supervision_edge_types, for negative edges by our function among all "non-existing" edge types in supervision_edge_types
>> also no valid/test edges in both directions

==================

if no negative edges sampling ratio (ns=0), the model predicts edges on all pairs of node (complete graph)! 
> So, the train and valid losses quickly become 0 as the best result and very early stop! 

<img width="1966" height="846" alt="Image" src="https://github.com/user-attachments/assets/f290de9c-6703-442c-baae-ede1fd864a9c" />

> but at test, the IR metrics become worst! 

<img width="652" height="789" alt="Image" src="https://github.com/user-attachments/assets/118d01a2-0dcc-4edb-a1ea-e100b696f0bb" />

**Comment by hosseinfani** [comment]
### For Node2Vec/MetaPath2Vec

#### During training:

They generate negatives by drawing uniformly random node sequences with no check against the actual edge set. This mirrors word2vec's original negative sampling: it doesn't verify that a "negative" context word/node isn't actually a true co-occurrence; it just accepts a small amount of noise as a tradeoff for cheap, fast sampling. 

#### During validation:

Seems possible ... I'm working on it

**Comment by hosseinfani** [comment]
Done for valid of n2v and m2v.

**Comment by hosseinfani** [comment]
For m2v during test, need to double check that the node ids are correctly mapped from global ids to m2v local ids.
